File: agent/src/main/java/pt/um/lei/masb/agent/SendMessages.java
package pt.um.lei.masb.agent;

import jade.content.lang.sl.SLCodec;
import jade.content.onto.BeanOntologyException;
import jade.core.behaviours.Behaviour;
import jade.domain.DFService;
import jade.domain.FIPAAgentManagement.DFAgentDescription;
import jade.domain.FIPAException;
import jade.lang.acl.ACLMessage;
import pt.um.lei.masb.agent.messaging.transaction.TransactionOntology;
import pt.um.lei.masb.blockchain.Transaction;
import pt.um.lei.masb.blockchain.utils.RingBuffer;

public class SendMessages extends Behaviour {
    private Transaction toSend;
    private RingBuffer<Transaction> rb;

    public SendMessages(RingBuffer<Transaction> tq){
        this.rb=tq;
    }


    @Override
    public void action() {
        DFAgentDescription dfd = new DFAgentDescription();
        try {
            DFAgentDescription[] agentList = DFService.search(myAgent, dfd);
            var codec = new SLCodec();
            TransactionOntology ontology= null;
            try {
                ontology = new TransactionOntology();
            } catch (BeanOntologyException e) {
                e.printStackTrace();
            }
            for (DFAgentDescription agent : agentList){
                for(Transaction t: rb) {
                    var msg = new ACLMessage(ACLMessage.INFORM);
                    //Transaction of the blockchain is not JADE serializable
                    //Need to convert into JTransaction of transaction ontology, in transaction.ontology package
                    //in order to actually send it.
                    myAgent.getContentManager().fillContent(msg,t);
                    msg.addReceiver(agent.getName());
                    msg.setLanguage(codec.getName());
                    msg.setOntology(ontology.getName());
                }
            }
        } catch (FIPAException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean done() {
        return false;
    }
}

File: agent/src/main/kotlin/pt/um/lei/masb/agent/behaviours/SendMessages.kt
package pt.um.lei.masb.agent.behaviours

import jade.content.lang.sl.SLCodec
import jade.core.behaviours.Behaviour
import jade.domain.FIPAException
import jade.lang.acl.ACLMessage
import mu.KLogging
import pt.um.lei.masb.agent.data.AgentPeers
import pt.um.lei.masb.agent.data.convertToJadeTransaction
import pt.um.lei.masb.agent.messaging.transaction.TransactionOntology
import pt.um.lei.masb.agent.messaging.transaction.ontology.actions.DiffuseTransaction
import pt.um.lei.masb.blockchain.SideChain
import pt.um.lei.masb.blockchain.Transaction
import pt.um.lei.masb.blockchain.data.BlockChainData
import pt.um.lei.masb.blockchain.utils.RingBuffer

class SendMessages(
    private val sc: SideChain,
    private val rb: RingBuffer<Transaction>,
    private val agentPeers: AgentPeers,
    private val clazz: Class<out BlockChainData>
//    private val srl: SerializationStrategy<T>
) : Behaviour() {
    private val toSend: Transaction? = null


    override fun action() {
        try {
            val codec = SLCodec()

            for (agent in agentPeers.ledgerPeers) {
                for (t in rb) {
                    val msg = ACLMessage(ACLMessage.INFORM)
                    myAgent.contentManager.fillContent(
                        msg,
                        DiffuseTransaction(
                            convertToJadeTransaction(
                                sc.blockChainId,
                                t
                            )
                        )
                    )
                    msg.addReceiver(agent)
                    msg.language = codec.name
                    msg.ontology = TransactionOntology.name
                }
            }
        } catch (e: FIPAException) {
            logger.error(e) { }
        }

    }

    override fun done(): Boolean {
        return false
    }

    companion object : KLogging()
}
